Migration of legacy config location files, and error logging, and constants

autoSteve opened the pull request:
Do you ever get the feeling that your forebears did not have OCD?

For ever since I first loaded this custom component it has made my eye twitch that its cache was stored in the base `config` location for Home Assistant.

Well, at that time there was only one. `solcast.json`. And that was acceptable if not a little triggering.

Our forebears did not consider bloating of use.

I have vastly contributed to much bloating on that front, with much cache, and much configuration being blasted there. It started with the sites cache. And then the usage cache. Then to granular dampening. Then crept out to undampened. And then swelled to actuals, dampened actuals, generation, and finally advanced options.

No more!

The sub-folder `solcast_solar` is becoming a thing. And migration is essential if this PR becomes main branch code.

This PR works well and is _entirely_ synthetically tested at this stage to 💯 on the `pytest` scale.

It is also flexible, and over-engineered with a constant to change behaviour and migration at will between 'to-subfolder' or 'to not subfolder', even cleaning up the subfolder if it becomes redundant.

And this code has not touched my prod. instance yet, which will be well backed up beforehand...

I would welcome comment from the likes of @Nilogax, who implemented a custom dampening solution, possibly naming things solcast*.json in the `config` folder. Those files would come along for the migration ride, possibly ruining someone's day.

Others might have things writing to `solcast-dampening.json` in the `config` folder, and I probably should monitor for that as well, moving updates into the new place.

Anyhoo, this is a "scorched earth" start, needing a `README.md` update, and probably more code to accommodate extant usage. The extant bits are going to likely only be granular dampening file-related only.

Nilogax commented:
Yes it always seemed strange that the json files sat in /config so I support the intention of this change. FWIW my original dampening solution kept itself to itself and didn't directly rely on any of the Solcast json files. Heck I even put the files it creates into their own subfolder under /config so maybe I was ahead of the game there. 

I have recently written a bit of code that retrieves actuals, forecast and generation for a day defined in a helper input date. That was the basis for my attempts at a damping efficacy calculation but it's no biggie to point that code somewhere else.  

I'd suggest anyone using the json files for their own porpoises will be 100% aware that they're doing it and competent enough to make the necessary changes.(as long as they read the release notes!). There's probably only a handful of us anyway.

autoSteve commented:
I like your thinking, and I like the warning even more. A a periodic monitor for ever is going to be a "meh" on a CPU/IO subsystem, but is a small load.

I shall think that one through fully. Great thought.

BJReplay commented:
> as long as they read the release notes!

Ha!

I suppose you believe in:

1. The tooth fairy
2. Santa Claus
3. Responsible Balanced Press

autoSteve commented:
> For ever?

This is done, @Nilogax.

Dead-man switch set for 1 June 2026 for file moves. The `pytest` gives up on testing moves at that date, too.

A warning is thrown. `Moved dampening file %s from legacy config to %s; Auto-moving will cease 1st June 2026`

As expected, adding an additional file monitor has added a rounding error to CPU utilisation. That said, I'm running on 2.2GHz Xeon Silvers, and definitely not a RPi. (I may or may not have _increased_ the frequency of file modification checks by 10x and still got negligible measurable difference, so I left it at 10th of a second. Legacy check languishes at half a second between checks, which is positively "yawn" by CPU standards.)

autoSteve commented:
A "smoking gun" for me is that the generation looks roughly half of forecast for every day... Maths will win.

```
[custom_components.solcast_solar.coordinator] Day 2025-10-25 forecasted 34.45 kWh vs actual 18.20 kWh
[custom_components.solcast_solar.coordinator] Day 2025-10-26 forecasted 19.52 kWh vs actual 8.25 kWh
[custom_components.solcast_solar.coordinator] Day 2025-10-27 forecasted 41.96 kWh vs actual 13.40 kWh
[custom_components.solcast_solar.coordinator] Day 2025-10-28 forecasted 47.74 kWh vs actual 23.05 kWh
[custom_components.solcast_solar.coordinator] Day 2025-10-29 forecasted 43.05 kWh vs actual 20.60 kWh
[custom_components.solcast_solar.coordinator] Day 2025-10-30 forecasted 46.02 kWh vs actual 21.85 kWh
[custom_components.solcast_solar.coordinator] Day 2025-10-31 forecasted 22.58 kWh vs actual 10.55 kWh
[custom_components.solcast_solar.coordinator] Day 2025-11-01 forecasted 42.42 kWh vs actual 21.75 kWh
[custom_components.solcast_solar.coordinator] Day 2025-11-02 forecasted 46.26 kWh vs actual 23.00 kWh
[custom_components.solcast_solar.coordinator] Day 2025-11-03 forecasted 21.36 kWh vs actual 6.15 kWh
[custom_components.solcast_solar.coordinator] Day 2025-11-04 forecasted 50.11 kWh vs actual 19.05 kWh
[custom_components.solcast_solar.coordinator] Day 2025-11-05 forecasted 54.07 kWh vs actual 23.30 kWh
[custom_components.solcast_solar.coordinator] Day 2025-11-06 forecasted 50.02 kWh vs actual 25.10 kWh
[custom_components.solcast_solar.coordinator] Day 2025-11-07 forecasted 39.12 kWh vs actual 12.05 kWh
[custom_components.solcast_solar.coordinator] Day 2025-10-26 forecasted 19.52 kWh vs actual 8.25 kWh
[custom_components.solcast_solar.coordinator] Day 2025-10-27 forecasted 42.71 kWh vs actual 13.40 kWh
[custom_components.solcast_solar.coordinator] Day 2025-10-28 forecasted 48.98 kWh vs actual 23.05 kWh
[custom_components.solcast_solar.coordinator] Day 2025-10-29 forecasted 43.05 kWh vs actual 20.60 kWh
[custom_components.solcast_solar.coordinator] Day 2025-10-30 forecasted 46.87 kWh vs actual 21.85 kWh
[custom_components.solcast_solar.coordinator] Day 2025-10-31 forecasted 22.58 kWh vs actual 10.55 kWh
[custom_components.solcast_solar.coordinator] Day 2025-11-01 forecasted 42.86 kWh vs actual 21.75 kWh
[custom_components.solcast_solar.coordinator] Day 2025-11-02 forecasted 46.89 kWh vs actual 23.00 kWh
[custom_components.solcast_solar.coordinator] Day 2025-11-03 forecasted 21.36 kWh vs actual 6.15 kWh
[custom_components.solcast_solar.coordinator] Day 2025-11-04 forecasted 50.20 kWh vs actual 19.05 kWh
[custom_components.solcast_solar.coordinator] Day 2025-11-05 forecasted 54.38 kWh vs actual 23.30 kWh
[custom_components.solcast_solar.coordinator] Day 2025-11-06 forecasted 51.00 kWh vs actual 25.10 kWh
[custom_components.solcast_solar.coordinator] Day 2025-11-07 forecasted 39.12 kWh vs actual 12.05 kWh
[custom_components.solcast_solar.coordinator] RMSE pv_estimate: 23.17 kWh, (24.05 kWh undampened)
[custom_components.solcast_solar.coordinator] RMSE pv_estimate10: 10.17 kWh, (10.93 kWh undampened)
[custom_components.solcast_solar.coordinator] RMSE pv_estimate90: 34.76 kWh, (35.34 kWh undampened)
```

autoSteve commented:
> how good Solcast's data is

It's really good for estimated actuals from what I've experienced. Good enough for me to bracket as "the truth" or I would not have based integrated automated dampening on its numbers. This feature relies on it being great.

That said it is not infallible, but it is bloody good for 'good' intervals of days, and that is all this feature needs. It craves 'good', and delta-logarithmically adjusts for 'bad'.

BJReplay commented:
> root mean squared error rate

Trying (but currently failing) to recall the gold standard error measure and rate we aspired to for electricity load forecast vs actuals reporting.

It will come to me.

As will the maths.

But it has been a decade since I had to do it in anger.

BJReplay commented:
> MAPE

MAPE it was.

> and backed away

I've told this story before; the abridged version is I spent two weeks in an office thinking and realised that I could get a forecasting process down from around 18 hours to around 6 minutes - not so much the forecasting, as calculating the actuals that went into the APE part of MAPE.  We needed a year's worth of actuals (1/2 hourly data) for all customers, compared to a year's worth of forecasts (same granularity), where each half hour had a different forecast price and actual price.  If the MAPE was good, the forecast (for the next day) was good.

The trick was that although each half hour had a different forecast price, it turned out that there were **lots** of customers who had the same combination of forecast prices across a year.  The process of finding that set of combinations was relatively quick, made quicker by signatures and caching on those forecasts, and so aggregating those customers together first dramatically reduced the size of the cartesian product.

autoSteve commented:
> so yesterday morning we built a dampening model

Given estimated actuals are usually pretty accurate I'd be building a RMSE or MAPE of dampened estimated actuals vs. generation, @Nilogax.

And this I have done. Kind of.

```
MAPE Forecasts...

[custom_components.solcast_solar.coordinator] MAPE calculation for pv_estimate: Day 2025-10-26 00:00:00, Actual 16.50 kWh, Estimate 19.52 kWh, Error 18.28%
[custom_components.solcast_solar.coordinator] MAPE calculation for pv_estimate: Day 2025-10-27 00:00:00, Actual 26.80 kWh, Estimate 41.96 kWh, Error 56.56%
[custom_components.solcast_solar.coordinator] MAPE calculation for pv_estimate: Day 2025-10-28 00:00:00, Actual 46.10 kWh, Estimate 47.74 kWh, Error 3.55%
[custom_components.solcast_solar.coordinator] MAPE calculation for pv_estimate: Day 2025-10-29 00:00:00, Actual 41.20 kWh, Estimate 43.05 kWh, Error 4.48%
[custom_components.solcast_solar.coordinator] MAPE calculation for pv_estimate: Day 2025-10-30 00:00:00, Actual 43.70 kWh, Estimate 46.02 kWh, Error 5.31%
[custom_components.solcast_solar.coordinator] MAPE calculation for pv_estimate: Day 2025-10-31 00:00:00, Actual 21.10 kWh, Estimate 22.58 kWh, Error 7.01%
[custom_components.solcast_solar.coordinator] MAPE calculation for pv_estimate: Day 2025-11-01 00:00:00, Actual 43.50 kWh, Estimate 42.42 kWh, Error 2.49%
[custom_components.solcast_solar.coordinator] MAPE calculation for pv_estimate: Day 2025-11-02 00:00:00, Actual 46.00 kWh, Estimate 46.26 kWh, Error 0.56%
[custom_components.solcast_solar.coordinator] MAPE calculation for pv_estimate: Day 2025-11-03 00:00:00, Actual 12.30 kWh, Estimate 21.36 kWh, Error 73.64%
[custom_components.solcast_solar.coordinator] MAPE calculation for pv_estimate: Day 2025-11-04 00:00:00, Actual 38.10 kWh, Estimate 50.11 kWh, Error 31.53%
[custom_components.solcast_solar.coordinator] MAPE calculation for pv_estimate: Day 2025-11-05 00:00:00, Actual 46.60 kWh, Estimate 54.07 kWh, Error 16.04%
[custom_components.solcast_solar.coordinator] MAPE calculation for pv_estimate: Day 2025-11-06 00:00:00, Actual 50.20 kWh, Estimate 50.02 kWh, Error 0.35%
[custom_components.solcast_solar.coordinator] MAPE calculation for pv_estimate: Day 2025-11-07 00:00:00, Actual 24.10 kWh, Estimate 39.12 kWh, Error 62.32%
[custom_components.solcast_solar.coordinator] MAPE calculation for pv_estimate: Day 2025-11-08 00:00:00, Actual 23.90 kWh, Estimate 23.15 kWh, Error 3.13%
[custom_components.solcast_solar.coordinator] MAPE calculation for pv_estimate: Day 2025-10-27 00:00:00, Actual 26.80 kWh, Estimate 42.71 kWh, Error 59.37%
[custom_components.solcast_solar.coordinator] MAPE calculation for pv_estimate: Day 2025-10-28 00:00:00, Actual 46.10 kWh, Estimate 48.98 kWh, Error 6.25%
[custom_components.solcast_solar.coordinator] MAPE calculation for pv_estimate: Day 2025-10-29 00:00:00, Actual 41.20 kWh, Estimate 43.05 kWh, Error 4.48%
[custom_components.solcast_solar.coordinator] MAPE calculation for pv_estimate: Day 2025-10-30 00:00:00, Actual 43.70 kWh, Estimate 46.87 kWh, Error 7.25%
[custom_components.solcast_solar.coordinator] MAPE calculation for pv_estimate: Day 2025-10-31 00:00:00, Actual 21.10 kWh, Estimate 22.58 kWh, Error 7.01%
[custom_components.solcast_solar.coordinator] MAPE calculation for pv_estimate: Day 2025-11-01 00:00:00, Actual 43.50 kWh, Estimate 42.86 kWh, Error 1.48%
[custom_components.solcast_solar.coordinator] MAPE calculation for pv_estimate: Day 2025-11-02 00:00:00, Actual 46.00 kWh, Estimate 46.89 kWh, Error 1.94%
[custom_components.solcast_solar.coordinator] MAPE calculation for pv_estimate: Day 2025-11-03 00:00:00, Actual 12.30 kWh, Estimate 21.36 kWh, Error 73.64%
[custom_components.solcast_solar.coordinator] MAPE calculation for pv_estimate: Day 2025-11-04 00:00:00, Actual 38.10 kWh, Estimate 50.20 kWh, Error 31.77%
[custom_components.solcast_solar.coordinator] MAPE calculation for pv_estimate: Day 2025-11-05 00:00:00, Actual 46.60 kWh, Estimate 54.38 kWh, Error 16.69%
[custom_components.solcast_solar.coordinator] MAPE calculation for pv_estimate: Day 2025-11-06 00:00:00, Actual 50.20 kWh, Estimate 51.00 kWh, Error 1.60%
[custom_components.solcast_solar.coordinator] MAPE calculation for pv_estimate: Day 2025-11-07 00:00:00, Actual 24.10 kWh, Estimate 39.12 kWh, Error 62.32%
[custom_components.solcast_solar.coordinator] MAPE calculation for pv_estimate: Day 2025-11-08 00:00:00, Actual 23.90 kWh, Estimate 23.15 kWh, Error 3.13%
[custom_components.solcast_solar.coordinator] MAPE pv_estimate: 20.38%, (21.30% undampened)
[custom_components.solcast_solar.coordinator] MAPE pv_estimate10: 25.12%, (23.36% undampened)
[custom_components.solcast_solar.coordinator] MAPE pv_estimate90: 58.63%, (57.79% undampened)

MAPE estimated actuals dampened...

[custom_components.solcast_solar.coordinator] MAPE calculation for pv_estimate: Day 2025-10-26 00:00:00, Actual 16.50 kWh, Estimate 14.85 kWh, Error 9.99%
[custom_components.solcast_solar.coordinator] MAPE calculation for pv_estimate: Day 2025-10-27 00:00:00, Actual 26.80 kWh, Estimate 28.31 kWh, Error 5.64%
[custom_components.solcast_solar.coordinator] MAPE calculation for pv_estimate: Day 2025-10-28 00:00:00, Actual 46.10 kWh, Estimate 47.42 kWh, Error 2.87%
[custom_components.solcast_solar.coordinator] MAPE calculation for pv_estimate: Day 2025-10-29 00:00:00, Actual 41.20 kWh, Estimate 44.11 kWh, Error 7.06%
[custom_components.solcast_solar.coordinator] MAPE calculation for pv_estimate: Day 2025-10-30 00:00:00, Actual 43.70 kWh, Estimate 46.84 kWh, Error 7.18%
[custom_components.solcast_solar.coordinator] MAPE calculation for pv_estimate: Day 2025-10-31 00:00:00, Actual 21.10 kWh, Estimate 20.73 kWh, Error 1.73%
[custom_components.solcast_solar.coordinator] MAPE calculation for pv_estimate: Day 2025-11-01 00:00:00, Actual 43.50 kWh, Estimate 42.41 kWh, Error 2.51%
[custom_components.solcast_solar.coordinator] MAPE calculation for pv_estimate: Day 2025-11-02 00:00:00, Actual 46.00 kWh, Estimate 48.78 kWh, Error 6.04%
[custom_components.solcast_solar.coordinator] MAPE calculation for pv_estimate: Day 2025-11-03 00:00:00, Actual 12.30 kWh, Estimate 10.36 kWh, Error 15.74%
[custom_components.solcast_solar.coordinator] MAPE calculation for pv_estimate: Day 2025-11-04 00:00:00, Actual 38.10 kWh, Estimate 44.35 kWh, Error 16.41%
[custom_components.solcast_solar.coordinator] MAPE calculation for pv_estimate: Day 2025-11-05 00:00:00, Actual 46.60 kWh, Estimate 51.67 kWh, Error 10.87%
[custom_components.solcast_solar.coordinator] MAPE calculation for pv_estimate: Day 2025-11-06 00:00:00, Actual 50.20 kWh, Estimate 52.71 kWh, Error 5.01%
[custom_components.solcast_solar.coordinator] MAPE calculation for pv_estimate: Day 2025-11-07 00:00:00, Actual 24.10 kWh, Estimate 21.78 kWh, Error 9.63%
[custom_components.solcast_solar.coordinator] MAPE calculation for pv_estimate: Day 2025-11-08 00:00:00, Actual 23.90 kWh, Estimate 23.84 kWh, Error 0.23%
[custom_components.solcast_solar.coordinator] MAPE calculation for pv_estimate: Day 2025-10-26 00:00:00, Actual 16.50 kWh, Estimate 14.81 kWh, Error 10.24%
[custom_components.solcast_solar.coordinator] MAPE calculation for pv_estimate: Day 2025-10-27 00:00:00, Actual 26.80 kWh, Estimate 28.32 kWh, Error 5.67%
[custom_components.solcast_solar.coordinator] MAPE calculation for pv_estimate: Day 2025-10-28 00:00:00, Actual 46.10 kWh, Estimate 48.70 kWh, Error 5.65%
[custom_components.solcast_solar.coordinator] MAPE calculation for pv_estimate: Day 2025-10-29 00:00:00, Actual 41.20 kWh, Estimate 44.13 kWh, Error 7.11%
[custom_components.solcast_solar.coordinator] MAPE calculation for pv_estimate: Day 2025-10-30 00:00:00, Actual 43.70 kWh, Estimate 48.03 kWh, Error 9.91%
[custom_components.solcast_solar.coordinator] MAPE calculation for pv_estimate: Day 2025-10-31 00:00:00, Actual 21.10 kWh, Estimate 21.02 kWh, Error 0.37%
[custom_components.solcast_solar.coordinator] MAPE calculation for pv_estimate: Day 2025-11-01 00:00:00, Actual 43.50 kWh, Estimate 42.83 kWh, Error 1.54%
[custom_components.solcast_solar.coordinator] MAPE calculation for pv_estimate: Day 2025-11-02 00:00:00, Actual 46.00 kWh, Estimate 49.86 kWh, Error 8.38%
[custom_components.solcast_solar.coordinator] MAPE calculation for pv_estimate: Day 2025-11-03 00:00:00, Actual 12.30 kWh, Estimate 10.32 kWh, Error 16.11%
[custom_components.solcast_solar.coordinator] MAPE calculation for pv_estimate: Day 2025-11-04 00:00:00, Actual 38.10 kWh, Estimate 44.41 kWh, Error 16.57%
[custom_components.solcast_solar.coordinator] MAPE calculation for pv_estimate: Day 2025-11-05 00:00:00, Actual 46.60 kWh, Estimate 52.07 kWh, Error 11.74%
[custom_components.solcast_solar.coordinator] MAPE calculation for pv_estimate: Day 2025-11-06 00:00:00, Actual 50.20 kWh, Estimate 53.99 kWh, Error 7.55%
[custom_components.solcast_solar.coordinator] MAPE calculation for pv_estimate: Day 2025-11-07 00:00:00, Actual 24.10 kWh, Estimate 22.00 kWh, Error 8.72%
[custom_components.solcast_solar.coordinator] MAPE calculation for pv_estimate: Day 2025-11-08 00:00:00, Actual 23.90 kWh, Estimate 23.84 kWh, Error 0.23%

[custom_components.solcast_solar.coordinator] MAPE pv_estimate: 7.21%, (7.84% undampened)

RMSE forecasts...

[custom_components.solcast_solar.coordinator] RMSE pv_estimate: 7.38 kWh, (7.82 kWh undampened)
[custom_components.solcast_solar.coordinator] RMSE pv_estimate10: 9.80 kWh, (9.76 kWh undampened)
[custom_components.solcast_solar.coordinator] RMSE pv_estimate90: 19.39 kWh, (20.04 kWh undampened)
```

And this tells us???

Yeah, applying dampening makes the forecasts/est. actuals just better on average. But not by much.

Unless my maths is completely broken the result is disappointing. But it is way less disppointing for individual MAPE estimated actuals, where some days were 'nailed', which makes me less disappointed. (Log a best/worst... 🤔)

This is a "rabbit hole" which I don't think I want to burrow further down in this minute. Simple `DEBUG` logging of the MAPE with respect to estimated actual pv_50 might be the furthest we want to go. Pleasingly the error rate is generally less than ten percent vs. dampened est. actuals, which is in line with what Solcast would like. I would like for more dampening going on to validate this further, but trees be gone at this time of year.

Amusingly, this mean absolute percentage error value almost exactly hits the root mean squared error calc of earlier for me.

autoSteve commented:
Quite possibly log a 10th/90th percentile of daily APE, or the classic 20th/80th rule. The MAPE overall for two weeks arguably. Maybe both of these.

And this would fall back in to @Nilogax territory of, "So... how accurate is Solcast? 🫣", but we know that Solcast is pretty good with estimated actuals, so that falls back to the territory of trust on that front.

In any event, this information would just give rope to those that want to say, "Ah ha! I told you your automated dampening numbers were bullshit!". To which I would reply, "Yeah? Go grope yourself. Roll your own solution per readme..."

And so (without vocal disagreement) this code is going to vanish. Ten... nine... eight... (going...) seven... (It's been fun, @Nilogax...) six...

autoSteve commented:
Somewhere in the readme now needs to mention _qu'ils mangent de la brioche_. The intent being, "Let them eat cake", even if the literal is anything but. Attributed to Marie Antoinette, but that is arguable.

I so want to work this in somehow, @Nilogax...

BJReplay commented:
Also: 

Since I'm fiddling, I did this:

``` json
{
    "forecast_day_entities": 14,
    "entity_logging": true,
    "reload_on_advanced_change": true
}
```

And, after enabling all the sensors, noticed this:

<img width="294" height="436" alt="image" src="https://github.com/user-attachments/assets/bd87d8db-fb31-4572-b946-792009129f3f" />

After a force update, saw this:

<img width="320" height="386" alt="image" src="https://github.com/user-attachments/assets/22c04441-a610-406d-ba92-b0022ee7c8dc" />

looking at code, suspect it might be this: https://github.com/BJReplay/ha-solcast-solar/blob/5755719df8cffb95766c45f631b09574695b43d1/custom_components/solcast_solar/coordinator.py#L126C58-L126C78 where `FORECAST_DAY_SENSORS` should probably be replaced with `coordinator.advanced_day_entities`

BJReplay commented:
> where `FORECAST_DAY_SENSORS` should probably be replaced with `coordinator.advanced_day_entities`

Make that `self.advanced_day_entities`

<img width="309" height="539" alt="image" src="https://github.com/user-attachments/assets/6a2e664e-6511-41d9-95b0-6190ab724085" />

autoSteve commented:
There is next to no `pytest` code for the _result_ of setting an advanced option. There is testing for the mechanics of reading the options and putting them in place.

This explains why the issue was never picked up.

Most of the advanced options just change what would have been a constant before, with the two options 'day sensor entities' and 'entity logging' being the most complex.

Enabling entity logging is really well tested because many test scenarios actually set this adv. opt. for real in the json file so that they can look for logged cues. I test the 'Solcast URL' well because I use a simulated API.

I think I'll add a test for the additional day forecast entities getting created and populated.

autoSteve commented:
Dampened estimated actual of `10.36`.

```
APE calculation for day 2025-11-03, Actual 12.30 kWh, Estimate 10.36 kWh, Error 15.74%
```

_**Undampened**_ estimated actual of `10.32`.

```
APE calculation for day 2025-11-03, Actual 12.30 kWh, Estimate 10.32 kWh, Error 16.11%
```

It should not be possible to have an undampened estimated actual total lower than the dampened estimated actual total for the same day.

What goes on here is that day intervals where export limiting has been detected (or forced) are excluded from both generation and estimated actual data sets to make them comparable. Yes, the day totals will appear lower for APE, but it's the percentage error rate that we're interested in so the actual values do not matter; it is the deviation of one compared to the other that we want.

So I will naval-gaze the impossible situation of undampened estimated actual being lower than dampened estimated actual. Something is NQR.

BJReplay commented:
> Get Split Enz tix?

Got distracted by the fact that I'm in Kyoto while TISM are playing Sydney Opera House.

No, not yet.

BJReplay commented:
> Get Split Enz tix?

Finally got around to it.

Also booked all Japan & Korea Accom.

Now need to sort out ferries (couple of islands on itinerary) and museums (a few will be very busy).

Just restarting picking up latest commits.

BJReplay commented:
``` text
2025-11-13 20:20:42.539 DEBUG (MainThread) [custom_components.solcast_solar.coordinator] Estimated actual MAPE: 10.57%, (9.60% dampened)
2025-11-13 20:20:42.539 DEBUG (MainThread) [custom_components.solcast_solar.coordinator] Estimated actual 50th percentile APE: 6.61%, (2.09% dampened)
2025-11-13 20:20:42.539 DEBUG (MainThread) [custom_components.solcast_solar.coordinator] Finished fetching solcast_solar data in 0.000 seconds (success: True)
```

autoSteve commented:
Y'all could. Or you could give me 15 minutes and I'll have the advanced option available. 😉😂

autoSteve commented:
Or ten minutes. Including doco, synthetic and real-world testing. Here you go.

BJReplay commented:
> Or ten minutes. Including doco, synthetic and real-world testing. Here you go.

``` text
2025-11-13 20:46:42.688 DEBUG (MainThread) [custom_components.solcast_solar.coordinator] Estimated actual mean APE: 10.57%, (9.60% dampened)
2025-11-13 20:46:42.688 DEBUG (MainThread) [custom_components.solcast_solar.coordinator] Estimated actual 10th percentile APE: 0.71%, (1.06% dampened)
2025-11-13 20:46:42.688 DEBUG (MainThread) [custom_components.solcast_solar.coordinator] Estimated actual 50th percentile APE: 6.61%, (2.09% dampened)
2025-11-13 20:46:42.689 DEBUG (MainThread) [custom_components.solcast_solar.coordinator] Finished fetching solcast_solar data in 0.000 seconds (success: True)
```

autoSteve commented:
Back under 4,000 statements now with a tiny refactor. 100% tested.

`solcastapi.py` needs a diet, because half of them come from that one module. デブ. That might need the end of year break to accomplish.

I'm thinking of breaking much supporting things out, but haven't got my head around what if any can be. Sub-classes? Much is intertwined that cannot be separated without careful thought and without making things even more complex. Even the date helper functions that need to know what the HA time zone is set as. Maybe convert a bunch to `@property` functions. A helper class for date things with `@property`? Probably. Dunno.

What I do know is that this code base is proving solid, with a very short list of question marks in my mind over things, mostly relating to summer/(winter, Irish ☘️🤦🏻) transition. I continue to mop those up.

It's just that `solcastapi.py` is super nubbly, and wants moving an arm and a leg and a beer gut to "repeatable", "reusable", "can-actually-find-without-autoSteve-involved" elsewhere else.

Big code is okay, don't get me wrong. But overwhelming classes with a massive number of massive functions is what trips folks up. `solcastapi.py` is on the edge.

It trips me up at times.

autoSteve commented:
> デブ

With the Katakana looking like a pair of droopy boobs followed by a round gut...

Jap translation, to save you going there: Syllabic, de-bu. Fat.

autoSteve commented:
Yesterday was a "good"/"brilliant" auto-dampening day...

(Remember that these kWh values exclude all intervals where site export has occurred, which was a lot of them yesterday afternoon.)

```
[custom_components.solcast_solar.coordinator] Calculating dampened estimated actual MAPE from 2025-11-01 to 2025-11-14
[custom_components.solcast_solar.coordinator] APE calculation for day 2025-11-14, Actual 6.40 kWh, Estimate 6.53 kWh, Error 1.97%
[custom_components.solcast_solar.coordinator] Calculating undampened estimated actual MAPE from 2025-11-01 to 2025-11-14
[custom_components.solcast_solar.coordinator] APE calculation for day 2025-11-14, Actual 6.40 kWh, Estimate 7.27 kWh, Error 13.61%
[custom_components.solcast_solar.coordinator] Estimated actual mean APE: 54.84%, (50.54% dampened)
[custom_components.solcast_solar.coordinator] Estimated actual 10th percentile APE: 9.38%, (1.61% dampened)
```

autoSteve commented:
Re-review. Heh-heh.

There's only 2,200 changed... 😬

BJReplay commented on custom_components/solcast_solar/solcastapi.py:
```diff
@@ -1797,16 +1868,16 @@ def get_rooftop_site_extra_data(self, site: str = "") -> dict[str, Any]:
             "name": _site.get("name"),
-            "resource_id": _site.get("resource_id"),
+            RESOURCE_ID: _site.get(RESOURCE_ID),
             "capacity": _site.get("capacity"),
```
So, I spot all the other strings replaced with CONSTs, and ask myself:
WHYNOTTHESE: _site.get(WHYNOTTHESE)?

BJReplay commented on custom_components/solcast_solar/coordinator.py:
```diff
@@ -814,39 +1062,39 @@ def get_sensor_extra_attributes(self, key: str = "") -> dict[str, Any] | None:
                         if factors.get(hour + ":00") is None:
-                            factors[hour + ":00"] = {"interval": hour + ":00", "factor": 1}
-                            factors[hour + ":30"] = {"interval": hour + ":30", "factor": 1}
+                            factors[hour + ":00"] = {"interval": hour + ":00", FACTOR: 1}
```
Another string that could be a const?

autoSteve commented:
> Does the renaming of the fetch method for the value from `get_remaining_today` to `forecast_remaining_today` cause a new sensor to created?

It must.

It's an interesting aspect of HA entities.

I think I should change this back.

BJReplay commented:
Unrelated: test run of solar excess EV charging yesterday (as we exited the doldrums of winter):

First chart - end of topping up Powerwall with grid power, then exporting solar.  Car plugged in just before 2pm.

Initial charge mode flat out (7kW), but switches to smart mode (excess solar) just before 2pm.

Second chart zooms in showing tracking of solar so that charge power generally utilises available solar, and allows for momentary excess household consumption to draw from Powerwall.

Temporary drops in charging power are when EV pauses charging to measure cell balance.

Offered charger power varied between 1380W (6A) and 2070W (9A).

<img width="2246" height="522" alt="image" src="https://github.com/user-attachments/assets/c65f7e26-4a5d-4b31-a781-2fc2fcfa8011" />

<img width="2278" height="497" alt="image" src="https://github.com/user-attachments/assets/bbb98a96-94a8-454f-9243-19c397200a59" />

autoSteve commented:
Ummm...

```
The cached data in /config/solcast_solar/solcast.json is corrupt in load_saved_data()
```

Yup... Zero length file. I have never before been able to replicate that.

```
-rw-r--r--    1 root     root             0 Nov 15 04:29 solcast.json
```

BJReplay commented:
> Not convinced about the `429`s.
> 
> I reverted, and a get sites fails for me

Same, and same.

Didn't stick around for the reversion zero length check.  Raised a support call.

BJReplay commented:
> Will go back and see if old logs have anything about solcast.json corruption.

No evidence of corruption here.

Failed start on reversion to 4.4.7 (as _now_ expected), clean (albeit) cached restart on pull of 4.4.8.

autoSteve commented:
I like that I take hourly backups. I'm back up and running.

Fresh commit to revert the entity naming for those errant entities. After restart you should be able to remove the `_2` versions and resume normal programming.

autoSteve commented:
The whole, "Let's zero the solcast.json file for fun" thing is fascinating me.

This happened when I reverted to 4.4.7, which goes nowhere near `/config/solcast_solar/solcast.json`, so _what the_? It must be an integration shutdown thing.

I know it happened at 04:29 UTC because I shared the console line here. I said I reverted at 04:33, so it was related to the reversion. It was after that when I realised the file was zero at 04:47 UTC. I grabbed the file from backup at 04:44 UTC.

It definitely happened on integration stop.

All logs are not gone.

An update that was failing with `429`s began at 04:26:50 UTC and in a 60-second pause. Hmmm. There is the smoking gun.

```
2025-11-15 15:28:46.218 WARNING (MainThread) [custom_components.solcast_solar.solcastapi] Call status 429/Try again later, pausing 60 seconds before retry
```

I cannot see the seconds on the file time, but now I have somewhere to stare at, and in the context of failure... I also have a way of recreating the situation.

That it only happens in a forecast retry situation is likely why it is a rare-as-rocking-horse-shit thing. I might raise an issue so folks have something to cling to and research their own situation if it occurs again. It's an obvious `WARNING`, so debug-level logs would not be required to work it out if I'm right.

BJReplay commented:
Yeah, I've added to my support call email, but so far just automatic replies.

Very annoying.

BJReplay commented:
You could (I'm betting in ~~five~~ ten minutes) make the UA an advanced option, and get 4.4.8 out the door :)

autoSteve commented:
> You could (I'm betting in ~five~ ten minutes) make the UA an advanced option, and get 4.4.8 out the door :)

An undocumented one smells right to me. 😅😎

autoSteve commented:
Well, you were kind-of right. Five minutes to code it, then ten minutes of furrowed brow confusion.

`429`s.

```
2025-11-16 11:31:08.124 DEBUG (MainThread) [custom_components.solcast_solar.solcastapi] Configuration directory is /config/solcast_solar
2025-11-16 11:31:08.125 DEBUG (MainThread) [custom_components.solcast_solar.solcastapi] Advanced options file /config/solcast_solar/solcast-advanced.json exists
2025-11-16 11:31:08.127 DEBUG (MainThread) [custom_components.solcast_solar.solcastapi] Advanced option set estimated_actuals_log_ape_percentiles: [10, 50]
2025-11-16 11:31:08.127 DEBUG (MainThread) [custom_components.solcast_solar.solcastapi] Advanced option set estimated_actuals_log_mape_breakdown: True
2025-11-16 11:31:08.127 DEBUG (MainThread) [custom_components.solcast_solar.solcastapi] Advanced option set reload_on_advanced_change: True
2025-11-16 11:31:08.127 DEBUG (MainThread) [custom_components.solcast_solar.solcastapi] Advanced option set user_agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/142.0.0.0 Safari/537.36
2025-11-16 11:31:08.128 DEBUG (MainThread) [custom_components.solcast_solar] Session headers {'accept': 'application/json', 'user_agent': 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/142.0.0.0 Safari/537.36'}
```

autoSteve commented:
It's not the code either. That's good. I tried the exact same hack as earlier and it failed this time. Ugh.

Committing the sleeper option anyway.

BJReplay commented:
> I have no answer for that.

No, about the only thing that we could do would be a status indicator.

Now...

You and I certainly could have an automation that fired a github action to enable a different issue.md when we are seeing a 429 storm.

Possibly even pin an announcement. 

I could do it most easily with an excess of API calls and set up a 24h/d schedule such that I spot it as early as possible.

Extending that, setting an observable indicator could then inform the installation config flow that a 429 storm was underway, but all of this seems like a **lot** of trouble.

autoSteve commented:
Just as an FYI, I am monitoring whether there is a possible export limiting bug.

Recall I mentioned that my MAPE had gone seemingly haywire? The reason was that all but the last couple of days of generation history had export limiting detected for every daylight hour. Impossible. This caused the error stats to be off-the-charts wrong, but also was defeating auto-dampen for me.

I can't see what could have caused this, but I keep looking.

To fix it for me I simply deleted the generation cache then re-started. I keep 14 days of history data in HA, so all history was restored as expected for the entire modelling period (because AdvancedOption™️), with export limiting detection for all intervals and days just fine. This tells me that a bug has not been recently introduced for get generation. That the last couple of days were fine before I renovated the entire data set that tells me that it may have been a bug that is now not a bug, given get generation usually only pulls the prior day.

A head-scratcher.

autoSteve commented:
This one's going to be a 'merge commit' and not a rebase. Should have done them the other way 'round. Ah, well.

Conflicts being resolved...

autoSteve commented:
(I could add an AdvancedOption™️ to fire an HA automation for you... 😂)

BJReplay commented:
> (I could add an AdvancedOption™️ to fire an HA automation for you... 😂)

Don't need that, I am pretty sure I can just use the persistent notification trigger

BJReplay commented:
> I'll get on to writing a couple of AdancedOption™️ things to fire an automation when 429 storm is detected and an issue is raised, and then once the issue is removed fire another one. This will align with when the integration \"repair\" issue gets created and goes away. Should ease things for you.

Too kind.

But more reliable than cobbled together pyscript.

autoSteve commented:
Got the little bugger.

It was in the function calculating forecast error.

autoSteve commented:
I can hold off releasing until later tomoz. I'm in no hurry. Unless the Internet is in a hurry to create yet another `429` storm...

BJReplay commented:
Entirely your call.
